File: src/main/java/org/lanternpowered/server/inventory/CarrierReference.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.inventory;

import org.spongepowered.api.block.tileentity.TileEntity;
import org.spongepowered.api.entity.Entity;
import org.spongepowered.api.item.inventory.Carrier;

import java.lang.ref.WeakReference;
import java.util.Optional;

import org.checkerframework.checker.nullness.qual.Nullable;

@SuppressWarnings("unchecked")
public interface CarrierReference<T> {

    static <T> CarrierReference<T> of(Class<T> carrierType) {
        return new CarrierReference<T>() {
            @Nullable private Object obj;

            @Override
            public Optional<Carrier> asCarrier() {
                return (Optional<Carrier>) get();
            }

            @Override
            public Optional<T> get() {
                if (this.obj == null) {
                    return Optional.empty();
                }
                if (this.obj instanceof WeakReference) {
                    return Optional.ofNullable(carrierType.cast(((WeakReference) this.obj).get()));
                }
                return Optional.of(carrierType.cast(this.obj));
            }

            @Override
            public <R> Optional<R> as(Class<R> ret) {
                return get().filter(ret::isInstance).map(ret::cast);
            }

            @Override
            public void set(@Nullable Carrier carrier) {
                this.obj = !carrierType.isInstance(carrier) ? null : (carrier instanceof TileEntity || carrier instanceof Entity) ?
                        new WeakReference<>(carrier) : carrier;
            }
        };
    }

    Optional<Carrier> asCarrier();

    Optional<T> get();

    <R> Optional<R> as(Class<R> ret);

    void set(@Nullable Carrier carrier);
}

File: src/main/kotlin/org/lanternpowered/server/inventory/carrier/CarrierReference.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.inventory.carrier

import org.lanternpowered.api.entity.Entity
import org.lanternpowered.api.util.uncheckedCast
import org.spongepowered.api.block.entity.BlockEntity
import org.spongepowered.api.item.inventory.Carrier
import java.lang.ref.WeakReference

interface CarrierReference<T : Any> {

    fun carrier(): Carrier?

    fun get(): T?

    fun <R> ofType(type: Class<R>): R?

    fun set(carrier: Carrier?)

    companion object {

        fun <T : Any> of(type: Class<T>): CarrierReference<T> = CarrierReferenceImpl(type)
    }
}

private class CarrierReferenceImpl<T : Any>(private val type: Class<T>) : CarrierReference<T> {

    private var value: Any? = null

    override fun carrier(): Carrier? =
            this.get() as Carrier?

    override fun get(): T? {
        val value = this.value ?: return null
        if (value is WeakReference<*>)
            return value.get().uncheckedCast()
        return value.uncheckedCast()
    }

    override fun <R> ofType(type: Class<R>): R? {
        val carrier = this.get()
        return if (type.isInstance(carrier)) carrier.uncheckedCast() else null
    }

    override fun set(carrier: Carrier?) {
        if (!this.type.isInstance(carrier)) {
            this.value = null
        } else if (carrier is Entity || carrier is BlockEntity) {
            this.value = WeakReference(carrier)
        } else {
            this.value = carrier
        }
    }
}
